Started with the custom status feature

manojmalik20 opened the pull request:
Added custom status actions and API integration
Added custom status selectors and types
Added custom status emoji component
Added EnableCustomStatuses flag in config
Added Set a status option in the settings sidebar

<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A brief description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) --> 

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs (for both iOS and Android if possible).
-->

manojmalik20 committed:
Started with the custom status feature
Added custom status actions and API integration
Added custom status selectors and types
Added custom status emoji component
Added EnableCustomStatuses flag in config
Added Set a status option in the settings sidebar

chetanyakan commented on app/components/sidebars/settings/settings_sidebar_base.js:
```diff
@@ -185,6 +189,33 @@ export default class SettingsSidebarBase extends PureComponent {
+        const labelComponent = isStatusSet ? <Text>{customStatus.text}</Text> : null;
+        const customStatusEmoji = (
+            <Emoji
+                emojiName={isStatusSet ? customStatus.emoji : 'trollface'}
```
We need to add a common emoji component instead of `trollface`

chetanyakan commented on app/components/custom_status/custom_status_emoji.tsx:
```diff
@@ -0,0 +1,33 @@
+    const customStatus = useSelector((state: GlobalState) => {
+        return getCustomStatus(state, userID);
+    });
+    if (!(customStatusEnabled && customStatus && customStatus.emoji)) {
```
```suggestion
    const emojiExists = customStatusEnabled && customStatus && customStatus.emoji;
    if (!emojiExists) {
```

manojmalik20 committed:
Fixed emoji and added clear button
Fixed the default emoji in the set a status in the settings sidebar
Made a custom status label component to match the styling of rest of the drawer items
Fixed the bug in the selectors
Added localization id in en.json

manojmalik20 committed:
Added icon for clear button
Fixed the custom status label issue
Added styling and icon for the clear button
Fixed order of the localization ids

manojmalik20 committed:
Made custom status modal with suggestions
Fixed the long custom status overflowing in the sidebar with ellipses
Added the functionality to open custom status modal from sidebar
Made the constants for custom status
Made the custom status suggestion functional component
Made the custom status modal functional component
Made the selector for the recent custom statuses
Added some localization ids

manojmalik20 committed:
Changed the custom status modal to class component
Added recent custom statuses in the modal
Fixed the styling of titles
Fixed lint errors except one

chetanyakan committed:
Fix types and use i18n translations

chetanyakan committed:
Update types

manojmalik20 committed:
Refactored some code and made clear button component
Refactored the code of the custom status label to a different file and used it everywhere
Made a separate component for clear button and added it in recents and sidebar
Fixed the styling of the custom status text according to the new figma designs

chetanyakan commented on app/components/sidebars/settings/drawer_item.js:
```diff
@@ -88,14 +90,19 @@ export default class DrawerItem extends PureComponent {
             >
                 <View style={style.container}>
                     {icon &&
```
```suggestion
                    {icon && (
```

chetanyakan commented on app/components/sidebars/settings/drawer_item.js:
```diff
@@ -88,14 +90,19 @@ export default class DrawerItem extends PureComponent {
+                        <View style={style.iconContainer}>
+                            {icon}
+                        </View>
                     }
```
```suggestion
                    )}
```

chetanyakan commented on app/components/sidebars/settings/drawer_item.js:
```diff
@@ -88,14 +90,19 @@ export default class DrawerItem extends PureComponent {
                         <View style={style.labelContainer}>
                             {label}
                         </View>
+                        {labelSibling &&
+                            <View style={style.labelSiblingContainer}>
+                                {labelSibling}
+                            </View>
+                        }
```
```suggestion
                        {labelSibling && (
                            <View style={style.labelSiblingContainer}>
                                {labelSibling}
                            </View>
                        )}
```

chetanyakan commented on app/components/sidebars/settings/settings_sidebar_base.js:
```diff
@@ -185,6 +204,60 @@ export default class SettingsSidebarBase extends PureComponent {
+                    style={{
+                        backgroundColor: changeOpacity(theme.centerChannelColor, 0.52),
+                        borderRadius: 1000,
+                    }}
```
nit - consider using a stylesheet instead

manojmalik20 committed:
Added emojipicker in custom status modal
Fixed the Done button in custom status modal
Added the functionality to clear status from inside the modal

hrishavkmr commented on app/screens/custom_status/custom_status_modal.tsx:
```diff
@@ -0,0 +1,315 @@
+        );
+    };
+
+    renderCustomStatusSuggestions = () => {
```
maybe we can use one renderSuggestions() to render recent as well as custom status suggestions.

manojmalik20 committed:
Merge pull request #2 from brightscout-alpha/MI-1190

Mi 1190
